### PR TITLE
[Dynamic Partition] reserve specific history periods by dynamic partition.

### LINKED ISSUE #6514: [DynamicPartition] support retention for specified period of history time.
**Is your feature request related to a problem? Please describe.**
In some situations, we need to keep partitions of some periods of time. To do so, we need to set `dynamic_partition.start` very big, and it will take lots of disk space.
For example, we need to keep partitions between `p20191031` to `p20191120`, now we should set `dynamic_partition.start = -730`, and in some days, they will be dropped day by day. In fact, we just need to keep 20+ partitions and never drop them.

**Describe the solution you'd like**
Add two properties of DynamicPartition. `dynamic_partition.reserved_history_starts` and `dynamic_partition.reserved_history_ends`. 
To support different periods of time, `dynamic_partition.reserved_history_starts` should be like `"2021-01-01, 2021-03-01, 2021-05-01, ..."` and correspondingly, `dynamic_partition.reserved_history_ends` should be like `"2021-01-31, 2021-03-11, 2021-05-31, ..."`. And the periods of time that is reserved should be `"2021-01-01" - "2021-01-31"`, `"2021-03-01" - "2021-03-31"`,  `"2021-05-01" - "2021-05-31"` and `...`.
And before dropping partition operation, we just need to exclude those partitions.

### PR DESCRIPTION
**Pull request by Henry2SS**
## Proposed changes

Add `RESERVED_HISTORY_STARTS` and `RESERVED_HISTORY_ENDS`. 
Fixes #6514 
## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ x] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Henry2SS** [comment]
I will add some unit-tests later. 
And I have already compiled and validated this function in my own clusters.

**Comment by morningman** [comment]
hi @Henry2SS please update the document

**Comment by Henry2SS** [comment]
> hi @Henry2SS please update the document

Done.

File: fe/fe-core/src/main/java/org/apache/doris/common/util/DynamicPartitionUtil.java [hunk 1]
```diff
@@ -222,6 +225,63 @@ private static void checkHotPartitionNum(String val) throws DdlException {
         }
     }
 
+    private static void checkReservedHistoryStarts(String reservedHistoryStarts) throws DdlException{
+        String[] starts = reservedHistoryStarts.split(",");
+        if (starts.length == 0) {
+            //ErrorReport.reportDdlException(ErrorCode.ERROR_DYNAMIC_PARTITION_RESERVED_HISTORY_STARTS_EMPTY);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/common/util/DynamicPartitionUtil.java:231]: Why not using `ErrorReport.reportDdlException`?

**Comment by Henry2SS** [comment]
Hi, @morningman changed reserved history periods from closedOpen to closed. Thanks for your advice.

**Comment by caiconghui** [comment]
how about comibine two configs into one config, use reserved_history_list instead of reserved_history_starts and reserved_history_ends, and we can set the property like "[2021-01-01, 2021-01-31] , [2021-02-03, 2021-02-15]", and use pattern to check the property and extract property

**Comment by Henry2SS** [comment]
> how about comibine two configs into one config, use reserved_history_list instead of reserved_history_starts and reserved_history_ends, and we can set the property like "[2021-01-01, 2021-01-31] , [2021-02-03, 2021-02-15]", and use pattern to check the property and extract property

Great idea. And i've finished changing it to `reserved_history_periods`. And ut will be modified later.

**Comment by caiconghui** [comment]
please rebase master to solve the conflicts problem

**Comment by Henry2SS** [comment]
> please rebase master to solve the conflicts problem

Done.

**Comment by Henry2SS** [comment]
@morningman Hi, it could reserve history periods when the TIME_UNIT is set to HOUR now. And modified docs and add two more unit-test to do validation.

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**